File: aviloo-mcp/src/aviloo_mcp/parser.py
"""Extract and parse AVILOO battery certificate documents (Swedish layout)."""

from __future__ import annotations

import re
import shutil
import subprocess
from pathlib import Path

_UUID_RE = re.compile(
    r"([0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12})"
)
_SOH_LABEL_RE = re.compile(
    r"H[\u00c4A]LSOTILLST[\u00c4A]ND\s*\(SOH\)[\s\S]{0,400}?(\d{1,3}(?:[.,]\d+)?)\s*%",
    re.IGNORECASE,
)
_PERCENT_RE = re.compile(r"(\d{1,3}(?:[.,]\d+)?)\s*%")
_ENERGY_RE = re.compile(
    r"(\d+(?:[.,]\d+)?)\s*kWh\s*\|\s*(\d+(?:[.,]\d+)?)\s*kWh",
    re.IGNORECASE,
)
_WLTP_RE = re.compile(
    r"(\d+(?:[.,]\d+)?)\s*km\s*\|\s*(\d+(?:[.,]\d+)?)\s*km",
    re.IGNORECASE,
)


def _pdftotext(pdf_path: Path) -> str:
    exe = shutil.which("pdftotext")
    if not exe:
        raise RuntimeError("pdftotext not found on PATH")
    result = subprocess.run(
        [exe, str(pdf_path), "-"],
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode != 0:
        raise RuntimeError(
            f"pdftotext failed ({result.returncode}): {result.stderr.strip()}"
        )
    return result.stdout


def _pypdf_text(pdf_path: Path) -> str:
    try:
        from pypdf import PdfReader
    except ImportError as exc:
        raise RuntimeError("pdftotext missing and pypdf not installed") from exc
    reader = PdfReader(str(pdf_path))
    parts: list[str] = []
    for page in reader.pages:
        parts.append(page.extract_text() or "")
    return chr(10).join(parts)


def extract_text(pdf_path: Path) -> str:
    """Extract plain text using pdftotext, optional pypdf fallback."""
    pdf_path = Path(pdf_path)
    if shutil.which("pdftotext"):
        return _pdftotext(pdf_path)
    return _pypdf_text(pdf_path)


def _parse_float(s: str) -> float:
    return float(s.replace(",", ".").replace(" ", ""))


def _parse_int_km(s: str) -> int:
    return int(re.sub(r"[^\d]", "", s))


def _field_after(text: str, label_pattern: str) -> str | None:
    m = re.search(label_pattern + r"\s*(.+)", text, re.IGNORECASE)
    if not m:
        return None
    line = m.group(1).strip()
    return line.split(chr(10))[0].strip() or None


def parse_aviloo_text(text: str, download_id: str | None = None) -> dict:
    """Parse AVILOO certificate fields from extracted text."""
    raw_soh_matches = [m.group(0) for m in _PERCENT_RE.finditer(text)]

    certificate_number: str | None = None
    cert_line = re.search(
        r"CERTIFIKATETS\s+NUMMER:\s*(" + _UUID_RE.pattern + r")",
        text,
        re.IGNORECASE,
    )
    if cert_line:
        certificate_number = cert_line.group(1).upper()
    else:
        m = _UUID_RE.search(text)
        if m:
            certificate_number = m.group(1).upper()

    brand = _field_after(text, r"VARUM[\u00c4A]RKE:")
    model = _field_after(text, r"MODELL:")
    vin = _field_after(text, r"VIN:")
    mileage_raw = _field_after(text, r"M[\u00c4A]TARST[\u00c4A]LLNING:")
    mileage_km = _parse_int_km(mileage_raw) if mileage_raw else None

    tested_at: str | None = None
    dt = re.search(
        r"DATUM\s+OCH\s+TID:\s*\n?\s*(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2})",
        text,
        re.IGNORECASE,
    )
    if dt:
        tested_at = dt.group(1).strip()

    performed_by = _field_after(text, r"UTF[\u00d6O]RD\s+AV:")

    soh_percent: float | None = None
    soh_m = _SOH_LABEL_RE.search(text)
    if soh_m:
        soh_percent = _parse_float(soh_m.group(1))
    else:
        for pm in _PERCENT_RE.finditer(text):
            val = _parse_float(pm.group(1))
            if 50 <= val <= 100:
                soh_percent = val
                break

    energy_current_kwh: float | None = None
    energy_new_kwh: float | None = None
    em = _ENERGY_RE.search(text)
    if em:
        energy_current_kwh = _parse_float(em.group(1))
        energy_new_kwh = _parse_float(em.group(2))

    wltp_current_km: float | None = None
    wltp_new_km: float | None = None
    wm = _WLTP_RE.search(text)
    if wm:
        wltp_current_km = _parse_float(wm.group(1))
        wltp_new_km = _parse_float(wm.group(2))

    assessment: str | None = None
    if re.search(r"UTM[\u00c4A]RKT\s+H[\u00c4A]LSA", text, re.IGNORECASE):
        assessment = "UTM\u00c4RKT H\u00c4LSA"
    certified = bool(
        re.search(r"AVILOO-certifierat", text, re.IGNORECASE)
        or re.search(r"officiellt\s+AVILOO", text, re.IGNORECASE)
    )

    return {
        "certificate_number": certificate_number,
        "download_id": download_id,
        "soh_percent": soh_percent,
        "vin": vin,
        "mileage_km": mileage_km,
        "tested_at": tested_at,
        "brand": brand,
        "model": model,
        "performed_by": performed_by,
        "assessment": assessment,
        "certified": certified,
        "energy_current_kwh": energy_current_kwh,
        "energy_new_kwh": energy_new_kwh,
        "wltp_current_km": wltp_current_km,
        "wltp_new_km": wltp_new_km,
        "raw_soh_matches": raw_soh_matches,
    }

File: internal/aviloo/parser/parser.go
package parser

import (
	"bytes"
	"fmt"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	pdf "github.com/ledongthuc/pdf"
)

var (
	uuidRE     = regexp.MustCompile(`([0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12})`)
	sohLabelRE = regexp.MustCompile(`(?i)H[\xC4A]LSOTILLST[\xC4A]ND\s*\(SOH\)[\s\S]{0,400}?(\d{1,3}(?:[.,]\d+)?)\s*%`)
	percentRE  = regexp.MustCompile(`(\d{1,3}(?:[.,]\d+)?)\s*%`)
	energyRE   = regexp.MustCompile(`(?i)(\d+(?:[.,]\d+)?)\s*kWh\s*\|\s*(\d+(?:[.,]\d+)?)\s*kWh`)
	wltpRE     = regexp.MustCompile(`(?i)(\d+(?:[.,]\d+)?)\s*km\s*\|\s*(\d+(?:[.,]\d+)?)\s*km`)
)

// Result holds parsed AVILOO certificate fields.
type Result map[string]any

// ExtractText reads plain text from a PDF using pdftotext when available, else a Go fallback.
func ExtractText(pdfPath string) (string, error) {
	if exe, err := exec.LookPath("pdftotext"); err == nil {
		out, err := exec.Command(exe, pdfPath, "-").Output()
		if err != nil {
			return "", fmt.Errorf("pdftotext failed: %w", err)
		}
		return string(out), nil
	}
	return pdfTextFallback(pdfPath)
}

func pdfTextFallback(pdfPath string) (string, error) {
	f, r, err := pdf.Open(pdfPath)
	if err != nil {
		return "", fmt.Errorf("pdftotext missing and pdf fallback failed: %w", err)
	}
	defer f.Close()
	var buf bytes.Buffer
	total := r.NumPage()
	for i := 1; i <= total; i++ {
		page := r.Page(i)
		if page.V.IsNull() {
			continue
		}
		text, err := page.GetPlainText(nil)
		if err != nil {
			continue
		}
		buf.WriteString(text)
		buf.WriteByte('\n')
	}
	return buf.String(), nil
}

func parseFloat(s string) float64 {
	s = strings.ReplaceAll(strings.ReplaceAll(s, ",", "."), " ", "")
	v, _ := strconv.ParseFloat(s, 64)
	return v
}

func parseIntKM(s string) int {
	digits := regexp.MustCompile(`\d`).FindAllString(s, -1)
	if len(digits) == 0 {
		return 0
	}
	v, _ := strconv.Atoi(strings.Join(digits, ""))
	return v
}

func fieldAfter(text, labelPattern string) *string {
	re := regexp.MustCompile("(?i)" + labelPattern + `\s*(.+)`)
	m := re.FindStringSubmatch(text)
	if len(m) < 2 {
		return nil
	}
	line := strings.TrimSpace(strings.Split(m[1], "\n")[0])
	if line == "" {
		return nil
	}
	return &line
}

// ParseAvilooText extracts certificate fields from extracted PDF text.
func ParseAvilooText(text string, downloadID *string) Result {
	rawSOH := make([]string, 0)
	for _, m := range percentRE.FindAllStringSubmatch(text, -1) {
		rawSOH = append(rawSOH, m[0])
	}

	var certificateNumber *string
	certLine := regexp.MustCompile(`(?i)CERTIFIKATETS\s+NUMMER:\s*(` + uuidRE.String() + `)`).FindStringSubmatch(text)
	if len(certLine) >= 2 {
		upper := strings.ToUpper(certLine[1])
		certificateNumber = &upper
	} else if m := uuidRE.FindStringSubmatch(text); len(m) >= 2 {
		upper := strings.ToUpper(m[1])
		certificateNumber = &upper
	}

	brand := fieldAfter(text, `VARUM[\xC4A]RKE:`)
	model := fieldAfter(text, `MODELL:`)
	vin := fieldAfter(text, `VIN:`)
	mileageRaw := fieldAfter(text, `M[\xC4A]TARST[\xC4A]LLNING:`)
	var mileageKM *int
	if mileageRaw != nil {
		v := parseIntKM(*mileageRaw)
		mileageKM = &v
	}

	var testedAt *string
	if dt := regexp.MustCompile(`(?i)DATUM\s+OCH\s+TID:\s*\n?\s*(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2})`).FindStringSubmatch(text); len(dt) >= 2 {
		v := strings.TrimSpace(dt[1])
		testedAt = &v
	}

	performedBy := fieldAfter(text, `UTF[\xD6O]RD\s+AV:`)

	var sohPercent *float64
	if sohM := sohLabelRE.FindStringSubmatch(text); len(sohM) >= 2 {
		v := parseFloat(sohM[1])
		sohPercent = &v
	} else {
		for _, pm := range percentRE.FindAllStringSubmatch(text, -1) {
			val := parseFloat(pm[1])
			if val >= 50 && val <= 100 {
				sohPercent = &val
				break
			}
		}
	}

	var energyCurrent, energyNew *float64
	if em := energyRE.FindStringSubmatch(text); len(em) >= 3 {
		c := parseFloat(em[1])
		n := parseFloat(em[2])
		energyCurrent = &c
		energyNew = &n
	}

	var wltpCurrent, wltpNew *float64
	if wm := wltpRE.FindStringSubmatch(text); len(wm) >= 3 {
		c := parseFloat(wm[1])
		n := parseFloat(wm[2])
		wltpCurrent = &c
		wltpNew = &n
	}

	var assessment *string
	if regexp.MustCompile(`(?i)UTM[\xC4A]RKT\s+H[\xC4A]LSA`).MatchString(text) {
		v := "UTMÄRKT HÄLSA"
		assessment = &v
	}

	certified := regexp.MustCompile(`(?i)AVILOO-certifierat`).MatchString(text) ||
		regexp.MustCompile(`(?i)officiellt\s+AVILOO`).MatchString(text)

	return Result{
		"certificate_number": certificateNumber,
		"download_id":        downloadID,
		"soh_percent":        sohPercent,
		"vin":                vin,
		"mileage_km":         mileageKM,
		"tested_at":          testedAt,
		"brand":              brand,
		"model":              model,
		"performed_by":       performedBy,
		"assessment":         assessment,
		"certified":          certified,
		"energy_current_kwh": energyCurrent,
		"energy_new_kwh":     energyNew,
		"wltp_current_km":    wltpCurrent,
		"wltp_new_km":        wltpNew,
		"raw_soh_matches":    rawSOH,
	}
}

// ParsePDF extracts and parses a PDF file.
func ParsePDF(pdfPath string) (Result, error) {
	text, err := ExtractText(pdfPath)
	if err != nil {
		return nil, err
	}
	stem := strings.TrimSuffix(filepath.Base(pdfPath), filepath.Ext(pdfPath))
	return ParseAvilooText(text, &stem), nil
}
